guides/inventing-service.md

```diff
@@ -0,0 +1,866 @@
+the `Service` trait Tower also provides several utilities that implement
+`Service` by wrapping some other type that also implements `Service`. Exactly
+like we did with `Timeout` and `JsonContentType`. Things can be composed in ways
+similar to what we've just did.
+
+Some example services provided by Tower:
+
+- `Timeout` - This is pretty much identical to the timeout we have built.
+- `Retry` - To automatically retry failed requests.
+- `RateLimit` - To limit the number of requests a service can handle over a
+  period of time.
+
+Types like `Timeout` and `JsonContentType` are typically called "middlewares"
+since they wrap another `Service` and transform the request or response in some
+way. Whereas types like `RequestHandler` is typically called "leaf services"
+since it sits at the leaves of a tree of nested services. The actual responses
+are normally produced in leaf services and modified by middlewares.
+
+The only thing left to talk about is "backpressure" and `poll_ready`.
+
+## Backpressure
+
+Imagine you wanted to write a rate limit middleware that wraps a `Service` and
+puts a limit on the maximum number of concurrent requests the underlying service
+can will receive. This would be useful if you had some service that had a hard
+upper limit on the amount of load it could handle.
+
+With our current `Service` trait we don't really have a good way to implement
+something like that. We could try:
+
+```rust
+impl<R, T> Service<R> for ConcurrencyLimit<T> {
+    fn call(&mut self, request: R) -> Self::Future {
+        // 1. Check a counter for the number of running requests
+        //    currently.
+        // 2. If there is capacity left send the request to `T`
+        //    and increment the counter.
+        // 3. If not somehow wait until capacity becomes available.
+    }
+}
+```
+
+If there is no capacity left we have to wait and somehow get notified when
+capacity becomes available. So if the service is receiving a lot of requests we
+risk spending a lot of time waiting. It would be nice if we had a way to tell
+the caller that there was no more capacity left. Maybe they wanted to react to
+that by calling another service or somehow creating another service to handle
+part of the load.
+
+In a way it would be nice if `Service` had a method like this:
+
+```rust
+trait Service<R> {
+    async fn ready(&mut self);
+}
+```
+
+`ready` would be an async function that completes when the service has capacity
+enough to receive one new request. We would then require users to first call
+`service.ready().await` before doing `service.call(request).await`.
+
+Separating "calling the service" from "reserving capacity" also unlocks new use
+cases such as being able to maintain a set of "ready services" that we keep up
+to date in the background such that when a request arrives we already have a
+ready service to send it to and don't have to first wait for it to become ready.
```

Comment:
or if its required to do a lot of work to say allocate the request you only need to do it when you can ensure it will get processed. (Think like reading a big image off the wire etc)